### ISSUE TITLE
Design document for z-layers

### ISSUE THREAD

**Issue by wence-** [thetisproject/thetis#22]
# Adding support for z-layers

In addition to the current support for sigma coordinates, we would like to be able to use z-layers.  This is important where we have shallow bathymetry and would like to exploit this to reduce the number of degrees of freedom we're using.  Also probably useful for wetting and drying.

The implementation will have cross-cutting implications across both Firedrake and PyOP2, but the aim here is to iron out the user-facing interface (and hence the necessary abstractions) which will guide the implementation.
## Variable layer columns

Currently, Firedrake only supports a single global number of extruded layers across the whole domain.  The first step is providing the ability to have a variable number of columns in the domain.  Proposed API:

``` python
base = Mesh(...)
start = MeshFunction(base, dtype=int)
stop = MeshFunction(base, dtype=int)
# The convenience version that just has a globally constant layer number remains.
mesh = ExtrudedMesh(base, start_layer=start, stop_layer=stop)
```

Supporting this should be reasonably straightforward.  Note that we do not expect to lose performance by making the number of layers per-column.  The performance advantage of extruded meshes comes from the data layout, not that we have hard-coded one of the loop bounds.
## The coordinate field

We would like not to have internal vertical facets that are "hanging".

```
+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
|   | <- This causes a problem
+---+
```

The reason for this is that we don't want to have to treat these specially in the code (all the logic becomes much more complicated).  The proposed solution is to create degenerate cells where there is a step in the number of layers, squashing the volume of the hanging facet to zero.

```
+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
|  /
+/
```

We will need to be a little careful here, because now all the terms in the Jacobian will be zero, and if we're don't take care, computations of J/detJ on the facet will create a NaN.
## Integrating on the vertical facets

I think the model is that "top" layer number is the same in all columns, and so the data model encodes the "start" layer in each column.  Since hanging vertical facets are squashed to zero, and therefore do not contribute non-zero values at any point, when we integrate over the vertical facets we have to inspect the number of layers in both of the columns and choose the starting layer corresponding to the shortest column.

**Comment by wence-** [comment]
My intention with this issue is to get the design plan for implementing z-layers down on paper.  We can then figure out the steps we need to take in PyOP2 and Firedrake to actually implement the necessary functionality.  Please correct and expand it where I have things wrong or incomplete.

**Comment by dham** [comment]
I think that if it's not too much work, we should allow the start and stop layer to vary. The flat top exception just stinks of one of those cases where some user will instantly come up with a use case.

**Comment by wence-** [comment]
OK, so I don't think it's too much work, however, I don't see what the interface looks like.   Would we, rather than passing a single array of the number of layers in each column, pass a 2d array which indicates the start and stop layer for each column.  But how is that different than implicitly assuming that the start layer is always zero, and the stop layer varies (or vice versa).

**Comment by dham** [comment]
So passing arrays is kind of funny in parallel. Ideally I guess this is a MeshFunction (well, two MeshFunctions) over the base mesh.

It is different from defaulting all start layers to zero because different cells end up adjacent.

**Comment by wence-** [comment]
OK, that makes sense.  A mesh function also makes more sense.

**Comment by stephankramer** [comment]
Sounds all good to me. So in 3D (for an extruded 2D basemesh), I presume the model is that we first see which vertical edges need squashing by looking at the shortest column surrounding a node in the base mesh?  Then vertical facets will have zero, one or two of its vertical edges squashed giving a normal quad, a triangle or a completely squashed facet and prismatic elements may end up as a pyramid or a tet.

I guess what we are talking about here is more like a topological structure for the layers with a capability for merging of layers at the bottom (and the top).  The actual depth of those layers is a separate issue (with z-layers being a specific choice). Even more general would be to allow merging at arbitrary depths within the column, but that would require inbetween squashed cells, and I don't see any immediate use of that.

**Comment by tkarna** [comment]
Looks good to me too. I agree that allowing variable start layer would be good, you'd need that for ice shelves for instance if someone wants to venture in that direction. Also Stephan's suggested merging of in-between elements might be useful, but all the applications that I can think of right now would probably be fine with just top/bottom shaved cells.

Another point is that the sloping bottom of the shaved elements may be an issue if it's very steep, and I think there are cases where you'd like to have the actual z mesh with new vertical boundary facets. But I don't think we need to take it into account for now, as (I believe) we are mostly focusing on coastal applications. In many cases you can also work around the slope problem by smoothing the bathymetry and/or refining the horizontal mesh.

**Comment by wence-** [comment]
So to support "interior" (in some sense) vertical boundary facets, we need to think a little harder about how to describe which facets are what to pyop2.  Stefan, if I understand correctly, your arbitrary depth merging would be something like:

```
+---+---+---+
|   |   |   |
+---+---+---+
|  /|\  |   |
+/  |  \+---+
|   |   |   |
+---+---+---+
```

What I've currently described is indeed how to deal with the topological structure (this needs to happen in PyOP2).  We will separately need to think about how to actually alter the geometry appropriately.

**Comment by wence-** [comment]
I actually did all this!  So I think in theory the infrastructure is there.

**Comment by tkarna** [comment]
Yes, we still lack support for it on the Thetis side. I think it mostly boils down to implementing an API for generating generic vertical meshes.
